scripts/histmakers/mw_with_mu_eta_pt.py

```diff
@@ -136,20 +138,69 @@ def build_graph(df, dataset):
     isQCDMC = dataset.group == "QCD"
     require_prompt = "tau" not in dataset.name # for muon GEN-matching
 
+    unfold = args.unfolding and dataset.name in ["WplusmunuPostVFP", "WminusmunuPostVFP"]
+
+    apply_theory_corr = args.theoryCorr and dataset.name in corr_helpers
+
+    cvh_helper = data_calibration_helper if dataset.is_data else mc_calibration_helper
+    jpsi_helper = data_jpsi_crctn_helper if dataset.is_data else mc_jpsi_crctn_helper
+
+
     if dataset.is_data:
         df = df.DefinePerSample("weight", "1.0")
     else:
         df = df.Define("weight", "std::copysign(1.0, genWeight)")
 
     weightsum = df.SumAndCount("weight")
 
-    df = df.Filter("HLT_IsoTkMu24 || HLT_IsoMu24")
-    #df = df.Filter("event % 2 == 1") # test with odd/even events
+    if unfold:
+        # gen level definitions
+        if args.genLevel == "preFSR":
+            df = theory_tools.define_prefsr_vars(df)
+            df = df.Define("ptGen", "chargeVgen < 0 ? genl.pt() : genlanti.pt()")   
+            df = df.Define("etaGen", "chargeVgen < 0 ? abs(genl.eta()) : abs(genlanti.eta())")
+        elif args.genLevel == "postFSR":
+            pdgId = -13 if "Wplusmunu" in dataset.name else 13
+            df = df.Define("postFSR_mu", f"GenPart_status == 1 && (GenPart_statusFlags & 1) && GenPart_pdgId == {pdgId}")
+            df = df.Define("postFSR_mu_idx", "ROOT::VecOps::ArgMax(GenPart_pt[postFSR_mu])")
+            df = df.Define("ptGen", "GenPart_pt[postFSR_mu][postFSR_mu_idx]")
+            df = df.Define("etaGen", "abs(GenPart_eta[postFSR_mu][postFSR_mu_idx])")
+
+        # add histograms before any selection
+        df_xnorm = df
+        df_xnorm = df_xnorm.DefinePerSample("exp_weight", "1.0")
+
+        df_xnorm = theory_tools.define_theory_weights_and_corrs(df_xnorm, dataset.name, corr_helpers, args)
+
+        df_xnorm = df_xnorm.DefinePerSample("count", "0.5")
+
+        xnorm_axes = [*unfolding_axes, differential.axis_xnorm]
+        xnorm_cols = [*unfolding_cols, "count"]
+        
+        results.append(df_xnorm.HistoBoost("xnorm", xnorm_axes, [*xnorm_cols, "nominal_weight"]))
 
-    apply_theory_corr = args.theoryCorr and dataset.name in corr_helpers
+        scale_axes = [*unfolding_axes, differential.axis_xnorm, axis_ptVgen, axis_chargeVgen]
+        scale_cols = [*unfolding_cols, "count", "ptVgen", "chargeVgen"]
 
-    cvh_helper = data_calibration_helper if dataset.is_data else mc_calibration_helper
-    jpsi_helper = data_jpsi_crctn_helper if dataset.is_data else mc_jpsi_crctn_helper
+        syst_tools.add_pdf_hists(results, df_xnorm, dataset.name, xnorm_axes, xnorm_cols, args.pdfs, base_name="xnorm")
```

Comment:
Maybe this sequence of stuff should be a function, I guess it's purely duplicated wrt the nominal df

Reply:
Good point. I made a new function "add_theory_hists" in wremnants/syst_tools that is now used for xnorm histograms and nominal histograms in the highPU histmakers for mw, wlike, and mz_dilepton